src/server/services/pipeline-in-progress/processData.js

```diff
@@ -0,0 +1,169 @@
+				}
+				else if (startTimestamp.isBefore(prevReading.endTimestamp)) {
+					if (isCumulative) {
+						readingOK = false;
+						errMsg = 'The reading start time is before the previous endTime and cumulative so OED cannot use this reading.'
+					}
+					else if (!isFirst(prevReading.endTimestamp)) {
+						//Only treat this as a warning since the readings may be sent in a different order.
+						errMsg = 'The current reading startTime is not after the previous reading\'s end time. This is expected on the first reading ever.';
+					}
+				}
+			}
+			if (readingOK && Math.abs(startTimestamp.diff(prevReading.endTimestamp)) > Tgap) {
+				if (isCumulative) {
+					readingOK = false;
+					errMsg = 'The end of the previous reading is too far from the start of the next readings in cumulative data so drop the readings.';
+				}
+				else if (!isFirst(prevReading.endTimestamp)) {
+					//Only treat this as a warning since we expect a gap in the first ever reading.
+					errMsg = 'There is a gap in time between this reading and the previous reading.';
+				}
+			}
+			if (readingOK) {
+				//This reading can be used
+				if (isCumulative) {
+					meterReading1 = rows[index - 2 * readingRepetition][0]; // use this if regular
+					meterReading2 = rows[index - readingRepetition][0];
+					meterReading = meterReading2 - meterReading1; // use this ifCumulative
+				}
+				else {
+					meterReading = rows[index - readingRepetition][0];
+				}
+				// Reject negative readings
+				if (isCumulative && meterReading1 < 0) {
+					// If the value isCumulative and the first reading is negative reject unless cumulativeReset is true
+					log.error(`DETECTED A NEGATIVE VALUE WHILE HANDLING CUMULATIVE READINGS FROM METER ${meterID}, ` +
+						`ROW ${index - readingRepetition}. REJECTED ALL READINGS`);
+					return [];
+				}
+				// To handle cumulative readings that resets at midnight
+				if (meterReading < 0) {
+					// if meterReading is negative and cumulative check that the times fall within an acceptable reset range
+					if (handleCumulativeReset(cumulativeReset, resetStart, resetEnd, startTimestamp)) {
+						meterReading = meterReading2;
+					}
+					else {
+						//cumulativeReset is not expected so throw an error
+						errMsg = ('A negative meterReading has been detected but either cumulativeReset is not enabled, or the start time and end time of this reading is out of the reset range.')
+						readingOK = false;
+					}
+				}
+				if (readingOK) {
+					if (Math.abs(prevReading.endTimestamp.diff(prevReading.startTimestamp) - endTimestamp.diff(startTimestamp)) 
+						> Tlen && !isFirst(prevReading.endTimestamp)) {
+						errMsg = 'The previous reading has a different time length than the current reading.';
+					}
+					currentReading = new Reading(meterID, meterReading, startTimestamp, endTimestamp);
+					result.push(currentReading);
+				}
+			}
+			if (!readingOK) {
```

Comment:
This is actually a little tricky because even though the final readingOK may be true it might be the case that the meterReading still errors out in which case the reading is dropped. 

So for example start times and end times might pass all checks and drop into if (readingOK). However, at this point we might find out that meterReading turns out to be negative and cumulative in which we case we must drop the reading and we still want to let the user know that the reading dropped even though it was ok up until this point. 

That's why I still think it's correct to leave it as an if !readingOK since if it occurs that readingOK is not true there is an accompanying error message that we should always let the client know,

HOWEVER, with the new changes to the meter model I've changed the logic so that meterReading is calculated along with start and end times in the beginning so it might make sense to move the checks out of the current if block and move it above the current if block. If we do this we can definitely do all the checks regarding meterReadings, start timestamps and end timestamps and then finally there are only two possibilities, either there is an error (readingOK == false )or no error (readingOK == true) in which case we can use an if else block. I hope this is clear but if it's not I can definitely try to make it clearer.